### PR TITLE
(PUP-8952) Respect server_list setting when using Rest::Client

### PR DESCRIPTION
**Pull request by Magisus**
This commit updates the server resolution logic of the
Puppet::Rest::Client to respect the `server_list` setting. Previously
is was falling back to the `server` setting (or whatever server it was
configured with) even when `server_list` was set and supposed to take
precedence.

This commit also updates the acceptance test to verify that
`server_list` is preferred first, to avoid hiding that large bug behind
less consequential failures.

### CODE REVIEW

**Comment by Magisus** [comment]
Thoughts on if there's a good central location to put this code? It currently differs only in that the original version did more aggressive nil checking, which I felt was unnecessary here considering that we are fairly clear about requiring the caller to provide settings. But if we wanted to reuse the code in both places, we could add that back in and move the methods to a central location, probably somewhere in Util?

**Comment by Magisus** [comment]
Decided to refactor the logic into a new `Util::Connection` module, but happy to move it anywhere.

**Comment by Magisus** [comment]
jenkins please test this

File: lib/puppet/indirector/rest.rb [hunk 1]
```diff
@@ -39,66 +40,19 @@ def self.srv_service
...
-          primary_server[0]
-        else
-          setting ||= :server
-          Puppet.debug "Dynamically-bound server lookup failed, falling back to #{setting} setting"
-          Puppet.settings[setting]
-        end
-      end
-      server
-    end
+    Puppet::Util::Connection.determine_server(setting)
```
Comment (amysahli) [inline lib/puppet/indirector/rest.rb]: Style suggestions: Lets maybe just put in server_setting directly here?  I think its just as clear
Puppet::Util::Connection.determine_server(server_setting)
But i'm happy either way

File: lib/puppet/indirector/rest.rb [hunk 2]
```diff
@@ -39,66 +40,19 @@ def self.srv_service
...
-          setting ||= :masterport
-          Puppet.debug "Dynamically-bound port lookup failed; falling back to #{setting} setting"
-          Puppet.settings[setting]
-        end
-      end
-      port.to_i
-    end
+    port_setting = port_setting()
+    server_setting = server_setting()
+    Puppet::Util::Connection.determine_port(port_setting, server_setting)
```
Comment (amysahli) [inline lib/puppet/indirector/rest.rb]: and same here? I think w/out definitions its just as clear
Puppet::Util::Connection.determine_port(port_setting, server_setting)

File: lib/puppet/rest/route.rb [hunk 3]
```diff
@@ -69,24 +72,8 @@ def with_base_url(dns_resolver)
...
-        bound_port = Puppet.lookup(:serverport) do
-          if primary_server = Puppet.settings[:server_list][0]
-            primary_server[1]
-          else
-            Puppet.settings[:masterport]
-          end
-        end
-
-        @server = bound_server
-        @port = bound_port
```
Comment (amysahli) [inline lib/puppet/rest/route.rb]: Won't @default_server && @default_port always be present now that we have the :server and :masterport defaults?
So do we still need the else statement?

File: lib/puppet/rest/route.rb [hunk 4]
```diff
@@ -69,24 +72,8 @@ def with_base_url(dns_resolver)
...
+        @server = determine_server(:server)
+        @port = determine_port(:masterport, :server)
```
Comment (amysahli) [inline lib/puppet/rest/route.rb]: If we do still need nil, lets use Puppet::Util::Connection.deternmine_stuff

File: spec/unit/rest/route_spec.rb [hunk 5]
```diff
@@ -45,10 +44,7 @@
 
       it 'yields the first entry in the server list when server_list is in use' do
         Puppet[:server_list] = [['one.net', 111], ['two.net', 222]]
-        fallback_route = Puppet::Rest::Route.new(api: '/fakeapi/v1/',
-                                                 default_server: nil,
-                                                 default_port: nil,
-                                                 srv_service: nil)
```
Comment (amysahli) [inline spec/unit/rest/route_spec.rb]: Lets keep one test where the user would specifcially pass in nil (so we know how things will work w/out the method defaults)

Comment (Magisus) [reply spec/unit/rest/route_spec.rb]: I've added a new test that explicitly passes `nil` for the settings.

File: lib/puppet/rest/route.rb (hunk 3 above)
Comment (Magisus) [reply lib/puppet/rest/route.rb]: You're right, I think we don't need any `if/else` here at all, it should be sufficient to just set `@server` and `@port` to the defaults here.

**Comment by Magisus** [comment]
Updated, thanks for checking the logic on this so thoroughly!

**Comment by amysahli** [comment]
jenkins please test this

**Comment by amysahli** [comment]
jenkins please test this

File: acceptance/tests/server_list_setting.rb [hunk 6]
```diff
@@ -6,57 +6,58 @@
...
+                       result.stdout, "should have selected the working master")
         end
+      end
+    end
+  end
+
+  step "Conflict warnings for server settings" do
+    with_puppet_running_on(master, {}) do
+      agents.each do |agent|
+        tmpconf = agent.tmpfile('puppet_conf_test')
```
Comment (amysahli) [inline acceptance/tests/server_list_setting.rb]: Looks like from the tests we need to move these lines (tmpconf=, etc) up above ```step "Server_list setting takes priority over server"``` so we can use "agent" and "tmpconf"

**Comment by Magisus** [comment]
I moved that test back inside the loop where it belongs, but still ahead of the other tests so we can see right off the bat that `server_list` is actually working.

**Comment by Magisus** [comment]
jenkins test this please

**Comment by Magisus** [comment]
jenkins please test this

**Pull request merged**